File: src/client/java/fr/hardel/asset_editor/client/javafx/routes/enchantment/EnchantmentSlotsPage.java
package fr.hardel.asset_editor.client.javafx.routes.enchantment;

import fr.hardel.asset_editor.client.javafx.VoxelColors;
import fr.hardel.asset_editor.client.javafx.VoxelFonts;
import fr.hardel.asset_editor.client.javafx.components.ui.Card;
import fr.hardel.asset_editor.client.javafx.components.ui.ResponsiveGrid;
import fr.hardel.asset_editor.client.javafx.components.ui.Section;
import fr.hardel.asset_editor.client.javafx.lib.RegistryPage;
import fr.hardel.asset_editor.client.javafx.lib.SlotManager;
import fr.hardel.asset_editor.client.javafx.lib.StudioContext;
import fr.hardel.asset_editor.client.javafx.lib.action.EnchantmentActions;
import fr.hardel.asset_editor.client.javafx.lib.data.SlotConfigs;
import fr.hardel.asset_editor.client.javafx.lib.data.SlotConfigs.SlotConfig;
import fr.hardel.asset_editor.client.javafx.lib.data.StudioBreakpoint;
import fr.hardel.asset_editor.network.EditorAction;
import javafx.geometry.Insets;
import javafx.scene.control.Label;
import javafx.scene.layout.VBox;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.core.registries.Registries;
import net.minecraft.world.entity.EquipmentSlotGroup;
import net.minecraft.world.item.enchantment.Enchantment;

import java.util.List;

public final class EnchantmentSlotsPage extends RegistryPage<Enchantment> {

    public EnchantmentSlotsPage(StudioContext context) {
        super(context, Registries.ENCHANTMENT, "enchantment-subpage-scroll", 16, new Insets(32));
    }

    @Override
    protected void buildContent() {
        Section section = new Section(I18n.get("enchantment:section.slots.description"));

        for (List<String> group : SlotConfigs.GROUPS) {
            ResponsiveGrid row = new ResponsiveGrid(ResponsiveGrid.autoFit(256))
                .atMost(StudioBreakpoint.XL, ResponsiveGrid.fixed(1));

            for (String slotId : group) {
                SlotConfig cfg = SlotConfigs.BY_ID.get(slotId);
                if (cfg == null)
                    continue;
                EquipmentSlotGroup slotGroup = EquipmentSlotGroup.valueOf(slotId.toUpperCase());
                Card card = new Card(cfg.image(), I18n.get("slot:" + cfg.id()), false);

                card.setOnMouseClicked(e -> applyAction(EnchantmentActions.toggleSlot(slotGroup),
                    new EditorAction.ToggleSlot(slotId)));
                bindView(entry -> new SlotManager(entry.data().definition().slots()).isActive(slotGroup),
                    card::setActive);

                row.addItem(card);
            }
            section.addContent(row);
        }

        section.addContent(buildExplanation());
        content().getChildren().setAll(section);
    }

    private VBox buildExplanation() {
        VBox box = new VBox(8);
        box.setPadding(new Insets(16));

        Label title = new Label(I18n.get("enchantment:slots.explanation.title"));
        title.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 13));
        title.setTextFill(VoxelColors.ZINC_300);
        title.setWrapText(true);

        Label item1 = new Label("- " + I18n.get("enchantment:slots.explanation.list.1"));
        Label item2 = new Label("- " + I18n.get("enchantment:slots.explanation.list.2"));
        for (Label item : new Label[] { item1, item2 }) {
            item.setFont(VoxelFonts.of(VoxelFonts.Variant.LIGHT, 13));
            item.setTextFill(VoxelColors.ZINC_400);
            item.setWrapText(true);
        }

        VBox list = new VBox(8, item1, item2);
        box.getChildren().addAll(title, list);
        return box;
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/routes/enchantment/EnchantmentSlotsPage.kt
package fr.hardel.asset_editor.client.compose.routes.enchantment

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import fr.hardel.asset_editor.client.compose.StudioBreakpoint
import fr.hardel.asset_editor.client.compose.StudioColors
import fr.hardel.asset_editor.client.compose.StudioTypography
import fr.hardel.asset_editor.client.compose.components.ui.*
import fr.hardel.asset_editor.client.compose.lib.*
import fr.hardel.asset_editor.client.memory.core.ClientWorkspaceRegistries
import fr.hardel.asset_editor.workspace.action.enchantment.ToggleSlotAction
import net.minecraft.client.resources.language.I18n

@Composable
fun EnchantmentSlotsPage(context: StudioContext) {
    val dialogs = rememberRegistryDialogState()
    val entry = rememberCurrentRegistryEntry(context, ClientWorkspaceRegistries.ENCHANTMENT) ?: return

    val activeSlots = remember(entry) {
        entry.data().definition().slots().map { it.serializedName }.toSet()
    }

    Column(
        verticalArrangement = Arrangement.spacedBy(32.dp),
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(32.dp)
    ) {
        Section(I18n.get("enchantment:section.slots.description")) {
            SlotConfigs.GROUPS.forEach { group ->
                ResponsiveGrid(
                    items = group.map { slotId ->
                        {
                            Card(
                                imageId = SlotConfigs.slotImage(slotId),
                                title = I18n.get("slot:$slotId"),
                                active = activeSlots.any { slot -> SlotConfigs.expandsTo(slot, slotId) },
                                onActiveChange = {
                                    context.dispatchRegistryAction(
                                        workspace = ClientWorkspaceRegistries.ENCHANTMENT,
                                        target = entry.id(),
                                        action = ToggleSlotAction(slotId),
                                        dialogs = dialogs
                                    )
                                }
                            )
                        }
                    },
                    defaultSpec = LayoutSpec.AutoFit(256.dp),
                    rules = listOf(BreakpointRule(maxWidth = StudioBreakpoint.XL.px.dp, spec = LayoutSpec.Fixed(floatArrayOf(1f))))
                )
            }

            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp),
                modifier = Modifier.padding(16.dp)
            ) {
                Text(
                    text = I18n.get("enchantment:slots.explanation.title"),
                    style = StudioTypography.regular(13),
                    color = StudioColors.Zinc300
                )
                Text(
                    text = "- ${I18n.get("enchantment:slots.explanation.list.1")}",
                    style = StudioTypography.light(13),
                    color = StudioColors.Zinc400
                )
                Text(
                    text = "- ${I18n.get("enchantment:slots.explanation.list.2")}",
                    style = StudioTypography.light(13),
                    color = StudioColors.Zinc400
                )
            }
        }
    }

    RegistryPageDialogs(context, dialogs)
}
